allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -189,130 +239,218 @@ def execute(self, logical_form: str) -> bool:
 
     def _execute_assertion(self, sub_expression: List) -> bool:
         """
-        Assertion functions are boolean functions. They take two arguments, which
-        evaluate to strings or integers, and compare them. The first element in the input list
-        should be the assertion function name, the second as an attribute function, and the
-        last element should be a constant.
-
-        Syntax: assertion_function(attribute_function, constant)
+        Assertion functions are boolean functions. They are of two types:
+        1) Exists functions: They take one argument, a set and check whether it is not empty.
+        Syntax: ``(exists_function function_returning_entities)``
+        Example: ``(object_exists (black (top all_objects)))`` ("There is a black object at the top
+        of a tower.")
+        2) Other assert functions: They take two arguments, which evaluate to strings or integers,
+        and compare them. The first element in the input list should be the assertion function name,
+        the second a function returning entities, and the last element should be a constant. The
+        assertion function should specify the entity type, the attribute being compared, and a
+        comparison operator, in that order separated by underscores. The following are the expected
+        values:
+            Entity types: ``object``, ``box``
+            Attributes being compared: ``color``, ``shape``, ``count``, ``color_count``,
+            ``shape_count``
+            Comparison operator:
+                Applicable to sets: ``all_equals``, ``any_equals``, ``none_equals``, ``same``,
+                ``different``
+                Applicable to counts: ``equals``, ``not_equals``, ``lesser``, ``lesser_equals``,
+                ``greater``, ``greater_equals``
+        Syntax: ``(assertion_function function_returning_entities constant)``
+        Example: ``(box_count_equals (member_shape_equals all_boxes shape_square) 2)``
+        ("There are exactly two boxes with only squares in them")
+
+        Note that the first kind is a special case of the second where the attribute type is
+        ``count``, comparison operator is ``greater_equals`` and the constant is ``1``.
         """
         # TODO(pradeep): We may want to change the order of arguments here to make decoding easier.
         assert isinstance(sub_expression, list), "Invalid assertion expression: %s" % sub_expression
         if len(sub_expression) == 1 and isinstance(sub_expression[0], list):
             return self._execute_assertion(sub_expression[0])
-        assert isinstance(sub_expression[0], str) and sub_expression[0].startswith("assert_"),\
+        is_assert_function = sub_expression[0].startswith("object_") or \
```

Comment:
This is a lot of complex logic.  Would it be simpler to just define a single function for each of your terminal functions, and call it?  Seems likely.  This also has the advantage of being much closer to the eventual use case of "you just write a python API, and we'll parse to that".

Reply:
Yeah, that should be possible. I'll add a TODO for this and come back to this later.